### PR TITLE
Migrate DataPayload to Yoke

### LINKED ISSUE #632: ICU4X data structs don't actually borrow any data
I discovered that even though we religiously use `Cow<'s, str>` and similar in our data structs, we are not actually benefiting from zero-copy deserialization, because the default implementation of `Deserialize` for `Cow<'s, str>` is to always use `Owned` data.  See a discussion here: https://github.com/serde-rs/serde/issues/1497

We need our structs to support zero-copy deserialization.  This requires adding `#[serde(borrow)]` to opt in where possible, keeping the following issues in mind: https://github.com/serde-rs/serde/issues/2016, https://github.com/serde-rs/serde/issues/723.

Unfortunately, if we make this change, it means that we will no longer be able to read from an `std::io::Read` when deserializing.  We must always load the buffer and then point into the buffer.  This *might* be an okay constraint for ICU4X, but I want to make sure we are explicit about it.

If we want to retain the ability to deserialize into a `'static`, we could add functions that flatten the lifetime on the data struct after we deserialize it from a borrowed buffer.

CC @Manishearth

### PR DESCRIPTION
**Pull request by sffc**
See #667

This migrates the internals of DataPayload to Yoke instead of Cow.

Outside of the core `icu_provider` crate, there are three types of changes:

1. On DataPayload types, a marker type suffixed with `_M` is used in place of the type `T`.
2. A second lifetime parameter needed to be added to DataProvider and related types.
3. Impls for related traits like Yokeable, ZeroCopyFrom, and DataMarker were added to the provider modules.

This change does *not* fix #632/#667 yet, but it lays the groundwork for that.

### CODE REVIEW

**Comment by sffc** [comment]
@Manishearth Here are my current type signatures (still with TODOs in some of the implementations)...

```rust
// FIXME: Change the name of this thing
pub trait DataStructHelperTrait {
    type Yokeable: for<'a> Yokeable<'a>;
}

enum DataPayloadInner<'d, T>
where
    T: DataStructHelperTrait,
{
    Borrowed(
        Yoke<
            <T as DataStructHelperTrait>::Yokeable,
            &'d <<T as DataStructHelperTrait>::Yokeable as Yokeable<'d>>::Output,
        >,
    ),
    RcStruct(
        Yoke<
            <T as DataStructHelperTrait>::Yokeable,
            Rc<<<T as DataStructHelperTrait>::Yokeable as Yokeable<'d>>::Output>,
        >,
    ),
    // Note: We could use either of the above two variants for Owned if we liked
    Owned(Yoke<<T as DataStructHelperTrait>::Yokeable, Option<&'static ()>>),
}

pub struct DataPayload<'d, T>
where
    T: DataStructHelperTrait,
{
    inner: DataPayloadInner<'d, T>,
}
```

The three variants support the following functions:

```rust
impl<'d, T> DataPayload<'d, T>
where
    T: DataStructHelperTrait,
{
    /// Convert a fully owned (`'static`) data struct into a DataPayload.
    #[inline]
    pub fn from_owned(data: <T as DataStructHelperTrait>::Yokeable) -> Self {
        // DataPayloadInner::Owned
    }

    /// Convert a partially owned (`'d`) data struct into a DataPayload.
    #[inline]
    pub fn from_partial_owned<'s: 'd>(
        data: <<T as DataStructHelperTrait>::Yokeable as Yokeable<'s>>::Output,
    ) -> Self {
        // DataPayloadInner::RcStruct
    }

    /// Convert a borrowed data struct into a DataPayload.
    #[inline]
    pub fn from_borrowed<'s: 'd>(
        data: &'d <<T as DataStructHelperTrait>::Yokeable as Yokeable<'s>>::Output,
    ) -> Self {
        // DataPayloadInner::Borrowed
    }
}
```

The good news is that the type signatures compile if I put `todo!()` as the implementation for those functions.  The bad news is that I'm having trouble implementing them.  I'd like if we could pair program a bit to figure it out.

**Comment by sffc** [comment]
@Manishearth I'm having trouble getting `attach_to_cart` and `ZeroCopyClone` to work nicely together.  The branch is in a state that should be useful for reproducing the error.

I think the problem is the signature of attach_to_cart:

```rust
    pub fn attach_to_cart_badly(
        cart: C,
        f: for<'de> fn(&'de <C as Deref>::Target) -> <Y as Yokeable<'de>>::Output,
    ) -> Self
```

There are no constraints on `<C as Deref>::Target`.  In my case, `<C as Deref>::Target` has its own lifetime parameter, and that lifetime parameter must exceed `'de`, but I don't know how to express that to the compiler.

**Comment by sffc** [comment]
@Manishearth This is ready for an initial review.  The changes are passing all tests in the `yoke` and `icu_provider` crates.  If those changes look good, I will proceed to migrate the rest of ICU4X, which will involve:

- Adding Yokeable, ZeroCopyClone, and DataMarker impls to all data structs
- Small tweaks around type signatures (e.g., adding the proposed `_M` suffix)

**Review by Manishearth — APPROVED** [review]

File: provider/core/src/serde.rs [hunk 1]
```diff
@@ -138,16 +139,87 @@ pub trait SerdeSeDataStruct<'s>: 's + Debug {
 
 impl_dyn_clone!(SerdeSeDataStruct<'s>, 's);
 
-impl_dyn_from_payload!(SerdeSeDataStruct<'s>, 'd, 's);
-
 impl<'s, T> SerdeSeDataStruct<'s> for T
 where
-    T: 's + serde::Serialize + Clone + Debug,
+    T: 's + serde::Serialize,
+    for<'a> &'a T: Clone,
```
[the remark this answers is not in the document]
Comment (sffc) [reply provider/core/src/serde.rs:145]: Follow-up: https://github.com/unicode-org/icu4x/issues/760

File: utils/yoke/src/zcc.rs [hunk 2]
```diff
@@ -0,0 +1,184 @@
...
+/// // Reference from a string slice directly
+/// impl ZeroCopyClone<str> for MyStruct<'static> {
+///     fn zcc<'b>(cart: &'b str) -> MyStruct<'b> {
+///         MyStruct {
+///             message: Cow::Borrowed(cart)
+///         }
+///     }
+/// }
+/// ```
+pub trait ZeroCopyClone<C: ?Sized>: for<'a> Yokeable<'a> {
```
[the remark this answers is not in the document]
Comment (sffc) [reply utils/yoke/src/zcc.rs]: Done

File: utils/yoke/src/zcc.rs [hunk 3]
```diff
@@ -0,0 +1,184 @@
...
+use std::borrow::Cow;
+use std::rc::Rc;
+
+/// Trait for types that can clone a cart type `C` with no allocations. A type can be the
+/// `ZeroCopyClone` target of multiple cart types.
+///
+/// It is possible to implement `ZeroCopyClone` without being zero-copy (using heap allocations),
+/// but doing so will reduce the performance of various `Yoke` constructors. The intention is for
+/// `ZeroCopyClone` to produce a struct from a cart with as little work as possible.
+///
```
[the remark this answers is not in the document]
Comment (sffc) [reply utils/yoke/src/zcc.rs]: Added some additional docs saying something like this.

File: provider/core/src/erased.rs [hunk 4]
```diff
@@ -56,79 +58,155 @@ pub trait ErasedDataStruct: 'static + Debug {
...
+                        }
+                        Err(_) => return Err(Error::MultipleReferences),
+                    },
+                    Err(any_rc) => any_rc,
+                };
+                let y2 = any_rc.downcast::<Yoke<M::Yokeable, Option<&'static ()>>>();
+                let any_rc = match y2 {
+                    Ok(rc_yoke) => match Rc::try_unwrap(rc_yoke) {
+                        Ok(yoke) => return Ok(DataPayload { inner: Owned(yoke) }),
+                        Err(_) => return Err(Error::MultipleReferences),
```
[the remark this answers is not in the document]
Comment (sffc) [reply provider/core/src/erased.rs:202]: Comment added.

File: provider/core/src/erased.rs [hunk 5]
```diff
@@ -56,79 +58,155 @@ pub trait ErasedDataStruct: 'static + Debug {
...
+        use crate::data_provider::DataPayloadInner::*;
+        match other.inner {
+            Borrowed(yoke) => {
+                // TODO(#752): This is not completely sound, because calling `.into_backing_cart()`
+                // throws away overrides stored in the Yokeable, such as those from `.with_mut()`.
+                let cart: &'d dyn ErasedDataStruct = yoke.into_backing_cart();
+                DataPayload::from_borrowed(cart)
+            }
+            RcStruct(yoke) => {
+                let cart: Rc<dyn ErasedDataStruct> = Rc::from(yoke);
```
[the remark this answers is not in the document]
Comment (sffc) [reply provider/core/src/erased.rs:102]: More comments added.

File: provider/core/src/data_provider.rs [hunk 6]
```diff
@@ -86,107 +91,215 @@ pub struct DataResponseMetadata {
...
+/// use icu_provider::marker::CowStr_M;
 ///
-/// let payload = DataPayload::from_borrowed("Demo");
+/// let payload = DataPayload::<CowStr_M>::from_borrowed("Demo");
 ///
 /// assert_eq!("Demo", payload.get());
 /// ```
-#[derive(Debug, Clone, PartialEq)]
-pub struct DataPayload<'d, T>
+pub struct DataPayload<'d, 's, M>
```
[the remark this answers is not in the document]
Comment (sffc) [reply provider/core/src/data_provider.rs:122]: Tracked already by https://github.com/unicode-org/icu4x/issues/621

File: components/plurals/src/provider.rs [hunk 7]
```diff
@@ -32,3 +32,29 @@ pub struct PluralRuleStringsV1<'s> {
     pub few: Option<Cow<'s, str>>,
     pub many: Option<Cow<'s, str>>,
 }
+
+// FIXME: Reconsider this
+pub struct PluralRuleStringsV1Helper {}
```
[the remark this answers is not in the document]
Comment (sffc) [reply components/plurals/src/provider.rs]: Changed to `PluralRuleStringsV1_M`

File: provider/core/src/data_provider.rs [hunk 8]
```diff
@@ -86,107 +91,215 @@ pub struct DataResponseMetadata {
...
+            Owned(yoke) => Owned(yoke.clone()),
+        };
         Self {
-            cow: Cow::Owned(data),
+            inner: new_inner
         }
     }
+}
+
+impl<'d, 's, M> PartialEq for DataPayload<'d, 's, M>
```
[the remark this answers is not in the document]
Comment (sffc) [reply provider/core/src/data_provider.rs:155]: I just deleted it.

File: provider/core/src/data_provider.rs [hunk 9]
```diff
@@ -141,15 +141,25 @@ where
 impl<'d, 's, M> Clone for DataPayload<'d, 's, M>
 where
     M: DataMarker<'s>,
+    for<'a> <M::Yokeable as Yokeable<'a>>::Output: Clone,
```
Comment (Manishearth) [inline provider/core/src/data_provider.rs:142]: Let's mark this (and the impl on DataResponse) with a comment linking to the rustc bug so folks don't get confused when this _doesn't_ work

File: provider/core/src/data_provider.rs [hunk 10]
```diff
@@ -160,7 +160,7 @@ where
     for<'a> &'a <M::Yokeable as Yokeable<'a>>::Output: PartialEq,
 {
     fn eq(&self, _other: &Self) -> bool {
-        todo!()
+        todo!("#753")
```
Comment (Manishearth) [inline provider/core/src/data_provider.rs]: suggestion (nb): Also perhaps add a full issue link in a comment and a short explanation?

**Review by Manishearth — APPROVED** [review]
looks great, minor non blocking suggestions

File: provider/core/src/data_provider.rs (hunk 10 above)
Comment (sffc) [reply provider/core/src/data_provider.rs]: This code is gone now.

File: provider/core/src/data_provider.rs (hunk 9 above)
Comment (sffc) [reply provider/core/src/data_provider.rs:142]: Added comment.

**Review by Manishearth — APPROVED** [review]

**Pull request merged**